File: agentproto/frame_test.go
package agentproto

import (
	"bytes"
	"encoding/hex"
	"encoding/json"
	"os"
	"strconv"
	"testing"
)

// frameEqual compares two frames, treating nil and empty Data as equal (Encode of
// an empty payload and DecodeFrame of a 1-byte frame differ only in nil-ness).
func frameEqual(a, b Frame) bool {
	return a.Op == b.Op && a.Rows == b.Rows && a.Cols == b.Cols && a.Seq == b.Seq && bytes.Equal(a.Data, b.Data)
}

func TestFrameRoundTrip(t *testing.T) {
	cases := []struct {
		name string
		in   Frame
	}{
		{"pty_out", PTYOutFrame([]byte("hello \x1b[31mworld\x1b[0m\r\n"))},
		{"pty_out_empty", PTYOutFrame(nil)},
		{"pty_out_binary", PTYOutFrame([]byte{0x00, 0x01, 0x02, 0xff, 0xfe})},
		{"input", InputFrame([]byte("ls -la\n"))},
		{"input_ctrl", InputFrame([]byte{0x03})}, // Ctrl-C
		{"resize", ResizeFrame(24, 80)},
		{"resize_zero", ResizeFrame(0, 0)},
		{"resize_max", ResizeFrame(65535, 65535)},
		{"hello_zero", HelloFrame(0)},
		{"hello", HelloFrame(4294967297)},
		{"hello_max", HelloFrame(18446744073709551615)},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got, err := DecodeFrame(tc.in.Encode())
			if err != nil {
				t.Fatalf("DecodeFrame(Encode()) error: %v", err)
			}
			if !frameEqual(got, tc.in) {
				t.Fatalf("round-trip mismatch:\n in  = %+v\n got = %+v", tc.in, got)
			}
		})
	}
}

func TestFrameEncodeWireBytes(t *testing.T) {
	// Opcode is the first byte; the wire layout is load-bearing (public contract).
	if b := PTYOutFrame([]byte("ab")).Encode(); !bytes.Equal(b, []byte{0x00, 'a', 'b'}) {
		t.Errorf("PTY_OUT wire bytes = % x", b)
	}
	if b := InputFrame([]byte("ab")).Encode(); !bytes.Equal(b, []byte{0x01, 'a', 'b'}) {
		t.Errorf("INPUT wire bytes = % x", b)
	}
	// RESIZE: opcode, then rows (big-endian uint16), then cols (big-endian uint16).
	if b := ResizeFrame(24, 80).Encode(); !bytes.Equal(b, []byte{0x02, 0x00, 24, 0x00, 80}) {
		t.Errorf("RESIZE wire bytes = % x", b)
	}
	if b := ResizeFrame(258, 259).Encode(); !bytes.Equal(b, []byte{0x02, 0x01, 0x02, 0x01, 0x03}) {
		t.Errorf("RESIZE wire bytes = % x", b)
	}
	// HELLO: opcode 0x04, then the start seq as a big-endian uint64.
	if b := HelloFrame(0).Encode(); !bytes.Equal(b, []byte{0x04, 0, 0, 0, 0, 0, 0, 0, 0}) {
		t.Errorf("HELLO(0) wire bytes = % x", b)
	}
	if b := HelloFrame(4294967297).Encode(); !bytes.Equal(b, []byte{0x04, 0, 0, 0, 1, 0, 0, 0, 1}) {
		t.Errorf("HELLO(2^32+1) wire bytes = % x", b)
	}
}

func TestDecodeFrameErrors(t *testing.T) {
	cases := []struct {
		name string
		raw  []byte
	}{
		{"empty", nil},
		{"unknown_opcode", []byte{0x7f, 'x'}},
		{"resize_too_short", []byte{byte(OpResize), 0x00, 24}},
		{"resize_too_long", []byte{byte(OpResize), 0, 24, 0, 80, 0}},
		{"hello_too_short", []byte{byte(OpHello), 0, 0, 0, 0, 0, 0, 1}},
		{"hello_too_long", []byte{byte(OpHello), 0, 0, 0, 0, 0, 0, 0, 1, 2}},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if _, err := DecodeFrame(tc.raw); err == nil {
				t.Fatalf("DecodeFrame(% x) = nil error, want error", tc.raw)
			}
		})
	}
}

func TestOpcodeString(t *testing.T) {
	for _, tc := range []struct {
		op   Opcode
		want string
	}{
		{OpPTYOut, "PTY_OUT"},
		{OpInput, "INPUT"},
		{OpResize, "RESIZE"},
		{OpRepaint, "REPAINT"},
		{OpHello, "HELLO"},
		{Opcode(0x42), "Opcode(0x42)"},
	} {
		if got := tc.op.String(); got != tc.want {
			t.Errorf("Opcode(%#x).String() = %q, want %q", byte(tc.op), got, tc.want)
		}
	}
}

// frameVector mirrors one entry of testdata/frame_vectors.json — the fixture the
// TypeScript codec (web/src/frame.test.ts) validates against too, so the browser
// and daemon codecs are pinned to byte-identical framing.
type frameVector struct {
	Name    string `json:"name"`
	Op      string `json:"op"`
	DataHex string `json:"dataHex"`
	Rows    uint16 `json:"rows"`
	Cols    uint16 `json:"cols"`
	Seq     string `json:"seq"` // decimal uint64 (string so JS BigInt keeps it exact)
	WireHex string `json:"wireHex"`
}

// frame reconstructs the logical Frame a vector describes.
func (v frameVector) frame(t *testing.T) Frame {
	t.Helper()
	switch v.Op {
	case "PTY_OUT":
		return PTYOutFrame(mustHex(t, v.DataHex))
	case "INPUT":
		return InputFrame(mustHex(t, v.DataHex))
	case "REPAINT":
		return RepaintFrame(mustHex(t, v.DataHex))
	case "RESIZE":
		return ResizeFrame(v.Rows, v.Cols)
	case "HELLO":
		seq, err := strconv.ParseUint(v.Seq, 10, 64)
		if err != nil {
			t.Fatalf("vector %q: bad seq %q: %v", v.Name, v.Seq, err)
		}
		return HelloFrame(seq)
	default:
		t.Fatalf("vector %q: unknown op %q", v.Name, v.Op)
		return Frame{}
	}
}

func mustHex(t *testing.T, s string) []byte {
	t.Helper()
	b, err := hex.DecodeString(s)
	if err != nil {
		t.Fatalf("bad hex %q: %v", s, err)
	}
	return b
}

// TestFrameGoldenVectors is the Go half of the cross-language codec contract: it
// asserts Encode produces the fixture's exact wire bytes and DecodeFrame reparses
// them back to the same Frame. web/src/frame.test.ts asserts the identical property
// against the SAME file, so the Go and TS codecs cannot silently diverge. On a
// mismatch this prints the actual bytes — the way to (re)derive a fixture entry.
func TestFrameGoldenVectors(t *testing.T) {
	raw, err := os.ReadFile("testdata/frame_vectors.json")
	if err != nil {
		t.Fatalf("read fixtures: %v", err)
	}
	var fixture struct {
		Vectors []frameVector `json:"vectors"`
	}
	if err := json.Unmarshal(raw, &fixture); err != nil {
		t.Fatalf("parse fixtures: %v", err)
	}
	if len(fixture.Vectors) == 0 {
		t.Fatal("no vectors in fixture")
	}
	for _, v := range fixture.Vectors {
		t.Run(v.Name, func(t *testing.T) {
			f := v.frame(t)
			wantWire := mustHex(t, v.WireHex)
			if got := f.Encode(); !bytes.Equal(got, wantWire) {
				t.Fatalf("Encode mismatch:\n got  = %x\n want = %s (%q)", got, v.WireHex, v.Name)
			}
			back, err := DecodeFrame(wantWire)
			if err != nil {
				t.Fatalf("DecodeFrame(%s) error: %v", v.WireHex, err)
			}
			if !frameEqual(back, f) {
				t.Fatalf("DecodeFrame round-trip mismatch:\n in  = %+v\n got = %+v", f, back)
			}
		})
	}
}

// TestStreamConsumerToleratesHello is the additive-safety property (#1592 Phase 5
// PR1): an existing stream consumer shaped exactly like the TUI/apiclient path
// (app/live_stream.go apiStream.Recv, integration ws_pty readPump) — which switches
// on Frame.Op and skips anything it doesn't handle — ignores the new OpHello frame
// with NO behavior change. OpHello decodes cleanly (it is a known op, so ReadMessage
// never errors), the consumer's switch falls through to its skip branch, no bytes are
// rendered, and the replay cursor is untouched. This is why the TUI is byte-for-byte
// unaffected by the broker now emitting a leading hello.
func TestStreamConsumerToleratesHello(t *testing.T) {
	// The exact wire sequence the broker now sends a fresh subscriber: hello, then a
	// repaint, then live output.
	wire := [][]byte{
		HelloFrame(42).Encode(),
		RepaintFrame([]byte("screen")).Encode(),
		PTYOutFrame([]byte("live")).Encode(),
	}
	// consume mirrors the TUI/apiclient frame switch: collect PTY_OUT bytes only,
	// render repaint without counting it, ignore everything else. A decode error
	// would model the consumer choking — which must NOT happen for OpHello.
	var out []byte
	var repainted bool
	for _, raw := range wire {
		f, err := DecodeFrame(raw)
		if err != nil {
			t.Fatalf("consumer decode error on % x: %v (a consumer would drop the stream)", raw, err)
		}
		switch f.Op {
		case OpPTYOut:
			out = append(out, f.Data...)
		case OpRepaint:
			repainted = true
		default:
			// INPUT/RESIZE/HELLO and any future op: skipped, exactly as apiStream.Recv
			// and the integration readPump do. The hello lands here — a no-op.
		}
	}
	if string(out) != "live" {
		t.Fatalf("PTY output after a leading hello = %q, want %q", out, "live")
	}
	if !repainted {
		t.Fatal("repaint frame was not delivered through the leading hello")
	}
}

File: web/src/frame.test.ts
// The TypeScript half of the cross-language codec contract (#1592 Phase 5 PR1).
// It validates web/src/frame.ts against the SAME fixture the Go test uses
// (agentproto/testdata/frame_vectors.json), so the browser and daemon codecs are
// pinned to byte-identical framing and cannot silently diverge. The Go half is
// agentproto/frame_test.go: TestFrameGoldenVectors — same file, same assertions.

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

import { Op, decode, encode, helloFrame, inputFrame, ptyOutFrame, repaintFrame, resizeFrame, type Frame } from "./frame.js";

const here = dirname(fileURLToPath(import.meta.url));
// web/src → repo-root/agentproto/testdata/frame_vectors.json (single source of truth).
const FIXTURE_PATH = join(here, "..", "..", "agentproto", "testdata", "frame_vectors.json");

interface Vector {
  name: string;
  op: "PTY_OUT" | "INPUT" | "RESIZE" | "REPAINT" | "HELLO";
  dataHex?: string;
  rows?: number;
  cols?: number;
  seq?: string; // decimal uint64 string — parsed with BigInt so it never truncates
  wireHex: string;
}

function fromHex(s: string): Uint8Array {
  if (s.length % 2 !== 0) throw new Error(`odd-length hex: ${s}`);
  const out = new Uint8Array(s.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(s.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

function toHex(b: Uint8Array): string {
  return Array.from(b, (x) => x.toString(16).padStart(2, "0")).join("");
}

function frameForVector(v: Vector): Frame {
  switch (v.op) {
    case "PTY_OUT":
      return ptyOutFrame(fromHex(v.dataHex ?? ""));
    case "INPUT":
      return inputFrame(fromHex(v.dataHex ?? ""));
    case "REPAINT":
      return repaintFrame(fromHex(v.dataHex ?? ""));
    case "RESIZE":
      return resizeFrame(v.rows ?? 0, v.cols ?? 0);
    case "HELLO":
      return helloFrame(BigInt(v.seq ?? "0"));
    default:
      throw new Error(`unknown op ${(v as Vector).op}`);
  }
}

function framesEqual(a: Frame, b: Frame): boolean {
  return (
    a.op === b.op &&
    a.rows === b.rows &&
    a.cols === b.cols &&
    a.seq === b.seq &&
    a.data.length === b.data.length &&
    a.data.every((x, i) => x === b.data[i])
  );
}

const fixture = JSON.parse(readFileSync(FIXTURE_PATH, "utf8")) as { vectors: Vector[] };

test("fixture has vectors", () => {
  assert.ok(fixture.vectors.length > 0, "no vectors in fixture");
});

for (const v of fixture.vectors) {
  test(`golden vector: ${v.name}`, () => {
    const f = frameForVector(v);

    // encode(f) must produce the fixture's exact wire bytes.
    assert.equal(toHex(encode(f)), v.wireHex, `encode mismatch for ${v.name}`);

    // decode(wire) must round-trip back to the same logical frame.
    const back = decode(fromHex(v.wireHex));
    assert.ok(framesEqual(back, f), `decode mismatch for ${v.name}: ${JSON.stringify({ f, back }, bigintReplacer)}`);
  });
}

// A couple of targeted OpHello assertions beyond the fixture loop, documenting the
// browser's reason for existing: the in-band start-seq it cannot read off the header.
test("hello carries the start seq as a big-endian uint64", () => {
  const wire = encode(helloFrame(4294967297n)); // 2^32 + 1
  assert.equal(toHex(wire), "040000000100000001");
  assert.equal(decode(wire).op, Op.Hello);
  assert.equal(decode(wire).seq, 4294967297n);
});

test("hello preserves the full uint64 range without truncation", () => {
  const max = 18446744073709551615n; // 2^64 - 1
  assert.equal(decode(encode(helloFrame(max))).seq, max);
});

test("decode rejects an unknown opcode (parity with Go DecodeFrame)", () => {
  assert.throws(() => decode(Uint8Array.from([0x7f, 0x78])), /unknown opcode/);
});

function bigintReplacer(_key: string, value: unknown): unknown {
  return typeof value === "bigint" ? value.toString() : value;
}
